File: Algorithms/0212_Word_Search_II/Python/Word_Search_II_Solution_1.py

# Space: O(n)
# Time: O(n)


import collections


class Solution():
    def findWords(self, board, words):
        column = len(board)
        if column == 0: return []
        row = len(board[0])
        if row == 0: return []

        target_word = set()
        res = set()

        # build a trie, this will be used in DFS search
        trie = collections.defaultdict(set)
        trie[''] = {word[0] for word in words}
        for each_word in words:
            string = ''
            for index in range(len(each_word)):
                string += each_word[index]
                if index != len(each_word) - 1:
                    trie[string].add(each_word[index + 1])
                else:
                    target_word.add(each_word)


        def search(word_list, cur_sub_string, x, y):
            if cur_sub_string in target_word:
                res.add(cur_sub_string)
                target_word.remove(cur_sub_string)

            if (not 0 <= x < row) or (not 0 <= y < column) or board[y][x] == 0 or board[y][x] not in trie[cur_sub_string] or len(res) == len(word_list): return

            cur_sub_string += board[y][x]
            temp = board[y][x]
            board[y][x] = 0
            search(word_list, cur_sub_string, x - 1, y)
            search(word_list, cur_sub_string, x + 1, y)
            search(word_list, cur_sub_string, x, y - 1)
            search(word_list, cur_sub_string, x, y + 1)
            board[y][x] = temp
            cur_sub_string = cur_sub_string[:-1]


        for y in range(column):
            for x in range(row):
                if len(words) == len(res):
                    return res
                search(words, '', x, y)

        return res

File: Algorithms/0212_Word_Search_II/Java/WordSearchIISolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    private int row;
    private int column;
    private List<String> res = new ArrayList<>();
    private Map<String, List<Character>> trie = new HashMap<>();
    private Set<String> targetWords = new HashSet<>();

    public List<String> findWords(char[][] board, String[] words) {
        column = board.length;
        row = board[0].length;
        for (String s : words) {
            trie.computeIfAbsent("", key -> new ArrayList<>()).add(s.charAt(0));
        }
        for (String word : words) {
            StringBuilder prefix = new StringBuilder("");
            for (int i = 0; i < word.length(); i++) {
                prefix.append(word.charAt(i));
                if (i < word.length() - 1) {
                    trie.computeIfAbsent(prefix.toString(), key -> new ArrayList<>()).add(word.charAt(i + 1));
                } else {
                    targetWords.add(prefix.toString());
                }
            }
        }
        for (int y = 0; y < column; y++) {
            for (int x = 0; x < row; x++) {
                dfs(board, new StringBuilder(), x, y, words);
            }
        }
        return res;
    }

    private void dfs(char[][] board, StringBuilder curStr, int x, int y, String[] words) {
        if (targetWords.contains(curStr.toString())) {
            res.add(curStr.toString());
            targetWords.remove(curStr.toString());
        }
        if (x < 0 || x >= row || y < 0 || y >= column || board[y][x] == '*' || !trie.getOrDefault(curStr.toString(), new ArrayList<>()).contains(board[y][x]) || words.length == res.size())
            return;
        // String temp = String.valueOf(board[y][x]);
        char temp = board[y][x];
        curStr.append(temp);
        board[y][x] = '*';
        dfs(board, curStr, x + 1, y, words);
        dfs(board, curStr, x - 1, y, words);
        dfs(board, curStr, x, y + 1, words);
        dfs(board, curStr, x, y - 1, words);
        board[y][x] = temp;
        curStr.deleteCharAt(curStr.length() - 1);
    }
}



